File: server/api/schemas/offer.py
from pydantic import BaseModel
from pydantic.utils import GetterDict
from typing import List

from .project import Project
from .user import User
from .distributor import Distributor
from .contactPerson import ContactPerson


class OfferItemGetter(GetterDict):
    def get(self, key: str, default = None):
        if key in {'id', 'name', 'model', 'category', 'quantity', 'comment', 'project'}:
            return getattr(self._obj.item, key)
        else:
            return super(OfferItemGetter, self).get(key, default)

class OfferItem(BaseModel):
    id: int
    name: str
    model: str
    category: str
    quantity: int
    comment: str
    project: Project
    price: float

    class Config:
        orm_mode = True
        getter_dict = OfferItemGetter

class OfferItemCreate(BaseModel):
    Item_id: int
    offer_id: int
    quantity: int
    price: float

class Offer(BaseModel):
    id: int
    user: User
    distributor: Distributor
    contactPerson: ContactPerson
    dateAndTime: str
    archived: bool
    items: List[OfferItem]

    class Config:
        orm_mode = True

class OfferCreate(BaseModel):
    idUser: int
    idDistributor: int
    idContactPerson: int
    archived: bool

class OfferEdit(BaseModel):
    id: int
    idUser: int
    idDistributor: int
    idContactPerson: int
    archived: bool

File: client/web/src/app/shared/models/offer.ts
import { IContactPerson, IDistributor, IItem, IUser } from ".";

export interface IPOffer {
    items?: IOffer[];
    previous_page?: number;
    next_page?: number;
    has_previous?: boolean;
    has_next?: boolean;
    total?: number;
    pages?: number;
}

export interface IOffer {
    id?: number;
    idUser?: number;
    user?: IUser;
    idDistributor?: number;
    distributor?: IDistributor;
    idContactPerson?: number;
    contactPerson?: IContactPerson;
    dateAndTime?: string;
    archived?: boolean;
    totalPrice?: number;
    items?: IOfferItem[];
}

export interface IOfferItem {
    Item_id?: number;
    offer_id?: number;
    quantity?: number;
    price?: number;
    currency?: string;
}
